### PR TITLE
Fix `process_execute_failed` only for supported new kernels

### LINKED ISSUE #3356: Missing symbols on newer kernels (> ~6.3)
## Description

I noticed that my 6.3.6 machine is missing the following symbols:
* `exec_binprm` - part of the `process_execute_failed` event
* `layout_and_allocate` - part of the `module_load` event

## Output of `tracee version`:

```
v0.16.1
```

## Output of `uname -a`:

```
Linux Alon-Zivony 6.3.6-microsoft-standard-WSL2 #1 SMP PREEMPT_DYNAMIC Mon Jun 12 11:59:43 IDT 2023 x86_64 x86_64 x86_64 GNU/Linux
```

### PR DESCRIPTION
**Pull request by AlonZivony**
### 1. Explain what the PR does

9fc0ae606 **fix(events): change process_execute_failed probes**
e9094a9c1 **feat(events): add probe relevance attribute**

Fix #3356

### CODE REVIEW

File: pkg/events/derive/process_execute_failed.go
Comment (NDStrahilevitz) [inline pkg/events/derive/process_execute_failed.go:48]: Why make this a singleton if it's not publicly accessible? Just construct the struct once outside the returned function and it will live in that scope.

File: pkg/events/definition_dependencies.go
Comment (NDStrahilevitz) [inline pkg/events/definition_dependencies.go:113]: nit: IsKernelCompatible

**Review by NDStrahilevitz — COMMENTED** [review]
Much better IMO. Thanks for addressing.

**Comment by NDStrahilevitz** [comment]
@josedonizetti or @geyslan I think one of you should review this as well due to the new kernel compatibility check.

File: pkg/events/core.go
Comment (AlonZivony) [reply pkg/events/core.go:12939]: So I added in the events the kernel compatibility to the probes

File: pkg/events/derive/process_execute_failed.go
Comment (AlonZivony) [reply pkg/events/derive/process_execute_failed.go:48]: Hmm its just not how it works right now with other functions.
I had initializations from the calling function with other derived events but they were removed.
Anyways, its more comftable API and I don't see a problem with doing it like this.

File: pkg/events/core.go
Comment (AlonZivony) [reply pkg/events/core.go:12939]: You want to add also kernel dependency to event dependencies?

**Comment by geyslan** [comment]
> @josedonizetti or @geyslan I think one of you should review this as well due to the new kernel compatibility check.

@AlonZivony when you move it for review, ping us. 👍🏼

File: pkg/events/derive/process_execute_failed.go
Comment (NDStrahilevitz) [reply pkg/events/derive/process_execute_failed.go:48]: If so, then it would also be redundant in them as well (unless they require public global access, not the case here as the singleton isn't exported or used in another internal file).

**Comment by AlonZivony** [comment]
> > @josedonizetti or @geyslan I think one of you should review this as well due to the new kernel compatibility check.
> 
> @AlonZivony when you move it for review, ping us. 👍🏼

This is ready for review :)

**Comment by geyslan** [comment]
> > > @josedonizetti or @geyslan I think one of you should review this as well due to the new kernel compatibility check.
> > 
> > 
> > @AlonZivony when you move it for review, ping us. 👍🏼
> 
> This is ready for review :)

Is this a fix, a feature or a chore? A bit of all? 😀 

If a fix, please open one issue to link it exposing the reasons. Tks.

I'm reviewing it now. 👍🏼

File: pkg/events/core.go
Comment (geyslan) [inline pkg/events/core.go:12895]: Could be a leftover?

Kernels:

- `Linux hb0 6.6.10-1-MANJARO #1 SMP PREEMPT_DYNAMIC Fri Jan  5 17:38:36 UTC 2024 x86_64 GNU/Linux`
- `Linux ip-172-31-5-229 5.4.0-137-generic #154-Ubuntu SMP Thu Jan 5 17:03:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`

`sudo ./dist/tracee -e process_execute_failed`

```json
{"level":"fatal","ts":1708628341.161495,"msg":"Tracee runner failed","error":"cmd.Runner.Run: error initializing Tracee: ebpf.(*Tracee).Init: ebpf.(*Tracee).initBPF: ebpf.(*Tracee).populateBPFMaps: failed to initialize tail call: ebpf.(*Tracee).initTailCall: could not get BPF program syscall__execveat__execute_failed: failed to find BPF program syscall__execveat__execute_failed: no such file or directory"}
```

**Review by geyslan — CHANGES_REQUESTED** [review]
LGTM when addressed a few things.

**Comment by geyslan** [comment]
> LGTM when addressed a few things.

@AlonZivony I have this PR #3848 that touches on some parts that you also touched on (e.g. attaching probe). I think it's worth taking a look just to catch on upcoming changes.

File: tests/e2e-inst-test.sh
Comment (AlonZivony) [reply tests/e2e-inst-test.sh:132]: So the comm is 16 bytes in the kernel, for a string.
So in practice, the comm received is 15 characters - hence the `process_execute`

File: tests/e2e-inst-signatures/scripts/process_execute_failed.sh
Comment (AlonZivony) [reply tests/e2e-inst-signatures/scripts/process_execute_failed.sh:11]: Just noticed that I get an error about the path executed if I don't use it.
I am not really sure why, but I just kept it.

File: pkg/events/core.go
Comment (AlonZivony) [reply pkg/events/core.go:12895]: You are right, I changed the name but didn't update it here.

Comment (AlonZivony) [reply pkg/events/core.go:12895]: It didn't pop up in the tests because I forgot to add the e2e to the `pr.yml` file...

**Comment by AlonZivony** [comment]
@geyslan I will give you a notice when the PR is ready again.
I have some issues with older kernels, I hope that I will soon make it work properly.

**Comment by AlonZivony** [comment]
@geyslan ok now it passes all the tests and ready for review :)

File: tests/e2e-inst-test.sh
Comment (geyslan) [inline tests/e2e-inst-test.sh:132]: Why tracee-ebpf is required here?

**Review by geyslan — APPROVED** [review]
LGTM

**Review by yanivagman — CHANGES_REQUESTED** [review]
Relying on kernel version to check if a symbol exists or not is not a good approach.
Commits can be backported (e.g. in RHEL), so we need a more robust way for these kind of checks.

**Comment by AlonZivony** [comment]
I opened #3983 to fix it quickly for now, as it is not using any choosing mechanism of probes.
When the fallback mechanism will be available we will be able to change the event to use it.

**Pull request closed without merging**
